File: tests/cpp/test_integer_optimization.cpp
/**
 * test_integer_optimization.cpp - Tests for integer parsing optimization
 *
 * Parity tests for tests/py/test_integer_optimization.py
 * These tests verify that integer parsing correctly handles all integer ranges.
 */

#include "strata/json/json_core.hpp"
#include "strata/json/json_parse.hpp"
#include "strata/json/json_serialize.hpp"

#include <cassert>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <limits>
#include <string>

using namespace strata;

static void assert_int_value(const JsonValue& value, int64_t expected) {
    assert(value.is_int());
    assert(value.as_int() == expected);
}

static void assert_double_value(const JsonValue& value, double expected) {
    assert(value.is_double());
    assert(value.as_double() == expected);
}

// ============================================================================
// Small Integer Parsing Tests
// ============================================================================

void test_zero() {
    auto result = parse_json("0");
    assert(result.ok());
    assert_int_value(result.value, 0);
    std::cout << "✓ test_zero passed\n";
}

void test_small_positive_integers() {
    auto r1 = parse_json("1");
    assert(r1.ok());
    assert_int_value(r1.value, 1);

    auto r2 = parse_json("127");
    assert(r2.ok());
    assert_int_value(r2.value, 127);

    auto r3 = parse_json("255");
    assert(r3.ok());
    assert_int_value(r3.value, 255);

    auto r4 = parse_json("256");
    assert(r4.ok());
    assert_int_value(r4.value, 256);

    std::cout << "✓ test_small_positive_integers passed\n";
}

void test_small_negative_integers() {
    auto r1 = parse_json("-1");
    assert(r1.ok());
    assert_int_value(r1.value, -1);

    auto r2 = parse_json("-5");
    assert(r2.ok());
    assert_int_value(r2.value, -5);

    auto r3 = parse_json("-128");
    assert(r3.ok());
    assert_int_value(r3.value, -128);

    std::cout << "✓ test_small_negative_integers passed\n";
}

void test_cache_boundary_values() {
    // Values at typical cache boundaries
    auto r1 = parse_json("-6");
    assert(r1.ok());
    assert_int_value(r1.value, -6);

    auto r2 = parse_json("257");
    assert(r2.ok());
    assert_int_value(r2.value, 257);

    std::cout << "✓ test_cache_boundary_values passed\n";
}

// ============================================================================
// Integer Boundary Tests
// ============================================================================

void test_int64_max() {
    // INT64_MAX = 2^63 - 1 = 9223372036854775807
    auto result = parse_json("9223372036854775807");
    assert(result.ok());
    assert_int_value(result.value, std::numeric_limits<int64_t>::max());
    std::cout << "✓ test_int64_max passed\n";
}

void test_int64_min() {
    // INT64_MIN = -2^63 = -9223372036854775808
    auto result = parse_json("-9223372036854775808");
    assert(result.ok());
    assert_int_value(result.value, std::numeric_limits<int64_t>::min());
    std::cout << "✓ test_int64_min passed\n";
}

void test_uint64_max() {
    // UINT64_MAX = 2^64 - 1 = 18446744073709551615
    auto result = parse_json("18446744073709551615");
    assert(result.ok());
    // Very large numbers may have precision loss as double
    double expected = 18446744073709551615.0;
    assert(result.value.is_double());
    assert(std::abs(result.value.as_double() - expected) / expected < 1e-15);
    std::cout << "✓ test_uint64_max passed\n";
}

void test_int32_boundaries() {
    // INT32_MAX = 2147483647
    auto r1 = parse_json("2147483647");
    assert(r1.ok());
    assert_int_value(r1.value, 2147483647);

    // INT32_MIN = -2147483648
    auto r2 = parse_json("-2147483648");
    assert(r2.ok());
    assert_int_value(r2.value, -2147483648LL);

    std::cout << "✓ test_int32_boundaries passed\n";
}

void test_long_boundaries() {
    // Test values at typical long boundaries
    auto r1 = parse_json("2147483647");
    assert(r1.ok());
    assert_int_value(r1.value, 2147483647);

    auto r2 = parse_json("-2147483648");
    assert(r2.ok());
    assert_int_value(r2.value, -2147483648LL);

    std::cout << "✓ test_long_boundaries passed\n";
}

// ============================================================================
// Integer Arrays and Objects Tests
// ============================================================================

void test_array_of_small_integers() {
    auto result = parse_json("[0, 1, 2, 3, 127, 255, 256]");
    assert(result.ok());
    const auto& arr = result.value.as_array();
    assert(arr.size() == 7);
    assert_int_value(arr[0], 0);
    assert_int_value(arr[1], 1);
    assert_int_value(arr[2], 2);
    assert_int_value(arr[3], 3);
    assert_int_value(arr[4], 127);
    assert_int_value(arr[5], 255);
    assert_int_value(arr[6], 256);
    std::cout << "✓ test_array_of_small_integers passed\n";
}

void test_array_of_negative_integers() {
    auto result = parse_json("[-1, -5, -6, -128, -256]");
    assert(result.ok());
    const auto& arr = result.value.as_array();
    assert(arr.size() == 5);
    assert_int_value(arr[0], -1);
    assert_int_value(arr[1], -5);
    assert_int_value(arr[2], -6);
    assert_int_value(arr[3], -128);
    assert_int_value(arr[4], -256);
    std::cout << "✓ test_array_of_negative_integers passed\n";
}

void test_array_of_mixed_integers() {
    auto result = parse_json("[-5, 0, 256, 257, 9223372036854775807]");
    assert(result.ok());
    const auto& arr = result.value.as_array();
    assert(arr.size() == 5);
    assert_int_value(arr[0], -5);
    assert_int_value(arr[1], 0);
    assert_int_value(arr[2], 256);
    assert_int_value(arr[3], 257);
    assert_int_value(arr[4], std::numeric_limits<int64_t>::max());
    std::cout << "✓ test_array_of_mixed_integers passed\n";
}

void test_object_with_integer_values() {
    auto result = parse_json("{\"zero\": 0, \"small\": 42, \"large\": 9223372036854775807}");
    assert(result.ok());
    const auto& obj = result.value.as_object();
    assert_int_value(obj.at("zero"), 0);
    assert_int_value(obj.at("small"), 42);
    assert_int_value(obj.at("large"), std::numeric_limits<int64_t>::max());
    std::cout << "✓ test_object_with_integer_values passed\n";
}

// ============================================================================
// Integer Round-trip Tests
// ============================================================================

void test_integer_roundtrip() {
    // Test various integer values through serialize and parse
    std::vector<int64_t> test_values = {
        0, 1, -1, -5, 256, 257, -6, 127, -128,
        2147483647LL, -2147483648LL
    };

    for (int64_t value : test_values) {
        std::string json_str = std::to_string(value);
        auto parsed = parse_json(json_str);
        assert(parsed.ok());
        assert_int_value(parsed.value, value);
    }

    std::cout << "✓ test_integer_roundtrip passed\n";
}

void test_serialize_roundtrip() {
    // Test serialization and re-parsing
    JsonValue::Object obj;
    obj["zero"] = JsonValue(JsonValue::Variant(static_cast<int64_t>(0)));
    obj["positive"] = JsonValue(JsonValue::Variant(static_cast<int64_t>(42)));
    obj["negative"] = JsonValue(JsonValue::Variant(static_cast<int64_t>(-123)));
    obj["large"] = JsonValue(JsonValue::Variant(static_cast<int64_t>(2147483647)));

    JsonValue root(JsonValue::Variant(std::move(obj)));
    std::string serialized = serialize_json(root);
    auto reparsed = parse_json(serialized);
    assert(reparsed.ok());
    assert_int_value(reparsed.value.as_object().at("zero"), 0);
    assert_int_value(reparsed.value.as_object().at("positive"), 42);
    assert_int_value(reparsed.value.as_object().at("negative"), -123);
    assert_int_value(reparsed.value.as_object().at("large"), 2147483647);

    std::cout << "✓ test_serialize_roundtrip passed\n";
}

// ============================================================================
// Edge Cases
// ============================================================================

void test_leading_zeros_rejected() {
    // Leading zeros are invalid in JSON numbers
    auto r1 = parse_json("01");
    assert(!r1.ok());

    auto r2 = parse_json("00");
    assert(!r2.ok());

    std::cout << "✓ test_leading_zeros_rejected passed\n";
}

void test_negative_zero() {
    // -0 should be parsed as 0
    auto result = parse_json("-0");
    assert(result.ok());
    assert_int_value(result.value, 0);
    std::cout << "✓ test_negative_zero passed\n";
}

void test_scientific_notation_integers() {
    // Scientific notation for integer values
    auto r1 = parse_json("1e2");
    assert(r1.ok());
    assert_double_value(r1.value, 100.0);

    auto r2 = parse_json("1E3");
    assert(r2.ok());
    assert_double_value(r2.value, 1000.0);

    std::cout << "✓ test_scientific_notation_integers passed\n";
}

// ============================================================================
// Main
// ============================================================================

int main() {
    std::cout << "Running Integer Optimization tests...\n\n";

    // Small integers
    std::cout << "--- Small Integer Parsing ---\n";
    test_zero();
    test_small_positive_integers();
    test_small_negative_integers();
    test_cache_boundary_values();

    // Integer boundaries
    std::cout << "\n--- Integer Boundaries ---\n";
    test_int64_max();
    test_int64_min();
    test_uint64_max();
    test_int32_boundaries();
    test_long_boundaries();

    // Arrays and objects
    std::cout << "\n--- Integer Arrays and Objects ---\n";
    test_array_of_small_integers();
    test_array_of_negative_integers();
    test_array_of_mixed_integers();
    test_object_with_integer_values();

    // Round-trip
    std::cout << "\n--- Integer Round-trip ---\n";
    test_integer_roundtrip();
    test_serialize_roundtrip();

    // Edge cases
    std::cout << "\n--- Edge Cases ---\n";
    test_leading_zeros_rejected();
    test_negative_zero();
    test_scientific_notation_integers();

    std::cout << "\n✅ All Integer Optimization tests passed!\n";
    return 0;
}

File: tests/py/test_integer_optimization.py
"""
Test cases for integer parsing optimization.

These tests verify that the optimized integer creation in PythonObjectBuilder
correctly handles all integer ranges, including:
- Small integers in Python's cache range (-5 to 256)
- Boundary values at cache edges
- Large integers requiring int64_t/uint64_t
- Edge cases at INT64_MIN, INT64_MAX, UINT64_MAX
"""

import sys
import pytest

import strata


class TestSmallIntegerParsing:
    """Verify small integers in Python's cached range are correctly parsed."""

    def test_zero(self):
        assert strata.loads("0") == 0

    def test_small_positive_integers(self):
        """Test positive integers in cache range."""
        assert strata.loads("1") == 1
        assert strata.loads("127") == 127
        assert strata.loads("255") == 255
        assert strata.loads("256") == 256

    def test_small_negative_integers(self):
        """Test negative integers in cache range."""
        assert strata.loads("-1") == -1
        assert strata.loads("-5") == -5
        assert strata.loads("-128") == -128

    def test_cache_boundary_values(self):
        """Test values at Python's small integer cache boundaries (-5 to 256)."""
        # Just below cache range
        assert strata.loads("-6") == -6
        # Just above cache range
        assert strata.loads("257") == 257


class TestIntegerBoundaries:
    """Test integer boundary handling for various C integer types."""

    def test_int64_max(self):
        """Test INT64_MAX (2^63 - 1)."""
        assert strata.loads("9223372036854775807") == 9223372036854775807

    def test_int64_min(self):
        """Test INT64_MIN (-2^63)."""
        assert strata.loads("-9223372036854775808") == -9223372036854775808

    def test_uint64_max(self):
        """Test UINT64_MAX (2^64 - 1).

        Note: Very large unsigned integers that exceed int64_t range may be
        parsed as floats by the underlying parser due to JSON number handling.
        This test verifies the approximate value is correct.
        """
        result = strata.loads("18446744073709551615")
        # The value may be returned as a float for numbers > INT64_MAX
        # Check approximate equality for float representation
        expected = 18446744073709551615
        if isinstance(result, float):
            # Float representation has limited precision for very large integers
            assert abs(result - expected) / expected < 1e-15
        else:
            assert result == expected

    def test_int32_boundaries(self):
        """Test INT32_MIN and INT32_MAX."""
        assert strata.loads("2147483647") == 2147483647  # INT32_MAX
        assert strata.loads("-2147483648") == -2147483648  # INT32_MIN

    def test_long_boundaries(self):
        """Test values at C long boundaries (platform dependent)."""
        # LONG_MAX is platform dependent (32-bit vs 64-bit)
        # On 64-bit systems, LONG_MAX == INT64_MAX
        # These values should always work regardless of platform
        assert strata.loads("2147483647") == 2147483647
        assert strata.loads("-2147483648") == -2147483648


class TestIntegerArraysAndObjects:
    """Test integer handling in arrays and objects."""

    def test_array_of_small_integers(self):
        """Array with integers in cache range."""
        result = strata.loads("[0, 1, 2, 3, 127, 255, 256]")
        assert result == [0, 1, 2, 3, 127, 255, 256]

    def test_array_of_negative_integers(self):
        """Array with negative integers."""
        result = strata.loads("[-1, -5, -6, -128, -256]")
        assert result == [-1, -5, -6, -128, -256]

    def test_array_of_mixed_integers(self):
        """Array with mixed small and large integers."""
        result = strata.loads("[-5, 0, 256, 257, 9223372036854775807]")
        assert result == [-5, 0, 256, 257, 9223372036854775807]

    def test_object_with_integer_values(self):
        """Object with integer values in various ranges."""
        result = strata.loads('{"zero": 0, "small": 42, "large": 9223372036854775807}')
        assert result["zero"] == 0
        assert result["small"] == 42
        assert result["large"] == 9223372036854775807


class TestIntegerTypePreservation:
    """Verify that integers are returned as Python int type."""

    def test_small_integer_is_int(self):
        result = strata.loads("42")
        assert isinstance(result, int)
        assert type(result) is int

    def test_large_integer_is_int(self):
        result = strata.loads("9223372036854775807")
        assert isinstance(result, int)
        assert type(result) is int

    def test_negative_integer_is_int(self):
        result = strata.loads("-42")
        assert isinstance(result, int)
        assert type(result) is int

    def test_zero_is_int(self):
        result = strata.loads("0")
        assert isinstance(result, int)
        assert type(result) is int


class TestIntegerRoundTrip:
    """Verify integer round-trip (loads/dumps) preserves values."""

    @pytest.mark.parametrize(
        "value",
        [
            0,
            1,
            -1,
            -5,
            256,
            257,
            -6,
            127,
            -128,
            2147483647,
            -2147483648,
            9223372036854775807,
            -9223372036854775808,
        ],
    )
    def test_integer_roundtrip(self, value):
        """Test that integers survive a loads/dumps round-trip."""
        json_str = str(value)
        parsed = strata.loads(json_str)
        assert parsed == value
        # Round-trip through dumps
        serialized = strata.dumps(parsed)
        reparsed = strata.loads(serialized)
        assert reparsed == value
